---
title: "CN-SNP"
output: html_document
---
```{r settings, message = FALSE, echo=FALSE}
library(knitr)
library(xtable)
library(dplyr)
library(ggplot2)
library(reshape2)
library(rsnps)
library(biomaRt)
library(NCBI2R)
library(magrittr)
library(XML)
library(RCurl)
# opts_chunk$set(dev=c("png"), 
#                dev.args=list(type="cairo"),
#                dpi =300, fig.height = 4, fig.width = 8)
# opts_chunk$set(fig_path = "/Users/DL/Documents/R%20project/Memphis%20-%20Copy%20Number%20Variation/Copy%20typing%20report/figure/")
# opts_chunk$set(dev=c("pdf"), 
#               fig.height =6, fig.width = 7)
opts_chunk$set(dev=c("svg"), 
              fig.height = 5, fig.width = 5)
# opts_chunk$set(dev="tiff", 
#                dev.args=list(compression="lzw"),
#                dpi=300)
opts_chunk$set(fig.show="hold")
opts_chunk$set(warning = F)
```

```{r my-funcitons}
getsnp <- function (SNPs) 
{
        SNPs = c("rs2066998", "rs376101" ,"rs397868160", "rs420358" )
#         print("succeed until here1!")
        tmp <- sapply(SNPs, function(x) grep("^rs[0-9]+$", x))
        if (tmp %>% sapply(length) %>% any == 0) {
                stop("not all items supplied are prefixed with 'rs';\n", 
                     "you must supply rs numbers and they should be prefixed with ", 
                     "'rs', e.g. rs420358")
        }
#         print("succeed until here2")
        query0 = "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=snp&mode=xml&id="
        query0 %>%
                paste(sep = "", paste(SNPs, collapse = ",")) %>%
                getURLContent() %>% 
                xmlInternalTreeParse() %>%
                xmlToList()  %>%
                .[1:length(.)-1] -> xml_list
#         print("succeed until here3")
        
        
        xml_list %>%
                sapply(., function(x) {
                        attr_rsIds <- tryget(x$.attrs["rsId"])
                        merge_indices <- which(names(x) == "MergeHistory")
                        if (length(merge_indices)) {
                               merge_rsIds <- sapply(x[merge_indices], "[[", "rsId")
                        }
                        else {
                                merge_rsIds <- NULL        
                        }
                        possible_names <- c(attr_rsIds, merge_rsIds)
                        return(possible_names)
                }) %>%
                unlist %>%
                unname %>% 
                (function(x) x[!is.na(x)]) %>%
                paste(sep = "", "rs", .) -> found_snps
        if(!(SNPs %in% found_snps) %>% any) {
                warning("The following rsIds had no information available on NCBI:\n  ", 
                        paste(SNPs[!(SNPs %in% found_snps)], collapse = ", "))
        }
                
        SNPs <- SNPs[SNPs %in% found_snps]
        
        my_freqs <- as.vector("numeric")
        for (i in seq_along(SNPs)) {
                xml_list[[i]]$Frequency["freq"] -> freqs
                my_freqs[i] <- ifelse (freqs %>% is.null, NA, freqs)
        }

        xml_list = xml_list[!my_freqs %>% is.na]
        SNPs = SNPs[!my_freqs %>% is.na]
 
        out <- as.data.frame(matrix(0, nrow = length(xml_list), ncol = 10))
        names(out) <- c("Query", "Chromosome", "Marker", "Class", 
                        "Gene", "Alleles", "Major", "Minor", "MAF", "BP")

        for (i in 1: length(xml_list)) {
                my_list = xml_list[[i]]
                my_chr <- xml_list[[i]] %>% (function(x) tryget(x$Assembly$Component$.attrs["chromosome"])) 
                if(is.null(my_chr)) {
                        my_chr <- NA
                        warning("No chromosomal information for", SNPs[i],
                                "; may be unmapped")
                }
                
                my_snp <- xml_list[[i]] %>% (function(x) tryget(x$.attrs["rsId"]))
                if(is.na(my_snp)) {
                        my_snp <- paste(sep = "", "rs", my_snp)
                }
                if (my_snp != SNPs[i]) {
                        warning(SNPs[i], " has been merged into ", my_snp)
                }
                
                
                my_snpClass <- xml_list[[i]] %>% (function(x) tryget(x$.attrs["snpClass"]))
                
                my_gene <- xml_list[[i]] %>% (function(x) tryget(x$Assembly$Component$MapLoc$FxnSet["symbol"]))
                if (is.null(my_gene)) 
                        my_gene <- NA
                
                alleles <- xml_list[[i]] %>% (function(x) tryget(x$Sequence$Observed))
                
                if (my_snpClass %in% c("snp", "snv", "in-del") & !is.null(xml_list[[i]]$Frequency)) {
                        tmp <- c(xml_list[[i]]$Ss$Sequence$Observed, xml_list[[i]]$Ss$.attrs["orient"])
                        if (tmp[2] != "forward") {
                                tmp[1] <- flip(tmp[1], sep = "/")
                        }
                        alleles_split <- unlist(strsplit(tmp[1], "/"))
                        maf_allele <- xml_list[[i]]$Frequency["allele"]
                        if (all(maf_allele != alleles_split)) {
                                maf_allele <- swap(maf_allele, c("A", "C", "G", 
                                                                 "T"), c("T", "G", "C", "A"))
                        }
                        
                 my_minor <- alleles_split[maf_allele == alleles_split]
                 
                 my_major <- alleles_split[maf_allele != alleles_split]
                 
                 my_freq <- my_list$Frequency["freq"]
#                         my_minor <- NA
#                         my_major <- NA
#                         my_freq <- NA
                }
                else {  
                        my_minor <- NA
                        my_major <- NA
                        my_freq <- NA
                }

                my_pos <- tryCatch(my_list$Assembly$Component$MapLoc$.attrs["physMapInt"], 
                                   error = function(e) {
                                           my_list$Assembly$Component$MapLoc["physMapInt"]
                                   })
                if (is.null(my_pos)) 
                        my_pos <- NA
                out[i, ] <- c(SNPs[i], as.integer(my_chr), my_snp, unname(my_snpClass), 
                              unname(my_gene), unname(alleles), unname(my_major), 
                              unname(my_minor), as.numeric(my_freq), as.integer(my_pos))
        }
        for (nm in c("MAF", "BP")) {
                out[, nm] <- as.numeric(out[, nm])
        }
        return(out)
        Sys.sleep(3)
}
```



```{r my-funcitons2}
getsnps <- function (SNPs) 
{
#         SNPs = c("rs2066998", "rs376101" ,"rs397868160", "rs420358" )
#         print("succeed until here1!")
        tmp <- sapply(SNPs, function(x) grep("^rs[0-9]+$", x))
        if (tmp %>% sapply(length) %>% any == 0) {
                stop("not all items supplied are prefixed with 'rs';\n", 
                     "you must supply rs numbers and they should be prefixed with ", 
                     "'rs', e.g. rs420358")
        }
#         print("succeed until here2")
        query0 = "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=snp&mode=xml&id="
        query0 %>%
                paste(sep = "", paste(SNPs, collapse = ",")) %>%
                getURL() %>% 
                xmlInternalTreeParse() %>%
                xmlToList()  %>%
                .[1:length(.)-1] -> xml_list
#         print("succeed until here3")
        
        
        xml_list %>%
                sapply(., function(x) {
                        attr_rsIds <- tryget(x$.attrs["rsId"])
                        merge_indices <- which(names(x) == "MergeHistory")
                        if (length(merge_indices)) {
                               merge_rsIds <- sapply(x[merge_indices], "[[", "rsId")
                        }
                        else {
                                merge_rsIds <- NULL        
                        }
                        possible_names <- c(attr_rsIds, merge_rsIds)
                        return(possible_names)
                }) %>%
                unlist %>%
                unname %>% 
                (function(x) x[!is.na(x)]) %>%
                paste(sep = "", "rs", .) -> found_snps
        if(!(SNPs %in% found_snps) %>% any) {
                warning("The following rsIds had no information available on NCBI:\n  ", 
                        paste(SNPs[!(SNPs %in% found_snps)], collapse = ", "))
        }
                
        SNPs <- SNPs[SNPs %in% found_snps]
        
        my_freqs <- as.vector("numeric")
        for (i in seq_along(SNPs)) {
                xml_list[[i]]$Frequency["freq"] -> freqs
                my_freqs[i] <- ifelse (freqs %>% is.null, NA, freqs)
        }

        xml_list = xml_list[!my_freqs %>% is.na]
        SNPs = SNPs[!my_freqs %>% is.na]
return(gsnp(SNPs))
}


```


```{r snp}
gsnp <- function (SNPs) 
{
        SNPs = SNPs[!is.na(SNPs)]        
    tmp <- sapply(SNPs, function(x) {
        grep("^rs[0-9]+$", x)
    })
    if (any(sapply(tmp, length) == 0)) {
        stop("not all items supplied are prefixed with 'rs';\n", 
            "you must supply rs numbers and they should be prefixed with ", 
            "'rs', e.g. rs420358")
    }
    query <- "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=snp&mode=xml&id="
    query <- paste(sep = "", query, paste(SNPs, collapse = ","))
    xml <- getURL(query)
    xml_parsed <- xmlInternalTreeParse(xml)
    xml_list_ <- xmlToList(xml_parsed)
    xml_list <- xml_list_[1:(length(xml_list_) - 1)]
    found_snps <- unname(unlist(sapply(xml_list, function(x) {
        attr_rsIds <- tryget(x$.attrs["rsId"])
        merge_indices <- which(names(x) == "MergeHistory")
        if (length(merge_indices)) {
            merge_rsIds <- sapply(x[merge_indices], "[[", "rsId")
        }
        else {
            merge_rsIds <- NULL
        }
        possible_names <- c(attr_rsIds, merge_rsIds)
        return(possible_names)
    })))
    found_snps <- found_snps[!is.na(found_snps)]
    found_snps <- paste(sep = "", "rs", found_snps)
    if (any(!(SNPs %in% found_snps))) {
        warning("The following rsIds had no information available on NCBI:\n  ", 
            paste(SNPs[!(SNPs %in% found_snps)], collapse = ", "))
    }
    SNPs <- SNPs[SNPs %in% found_snps]
    out <- as.data.frame(matrix(0, nrow = length(SNPs), ncol = 10))
    names(out) <- c("Query", "Chromosome", "Marker", "Class", 
        "Gene", "Alleles", "Major", "Minor", "MAF", "BP")
    for (i in seq_along(SNPs)) {
        my_list <- xml_list[[i]]
        my_chr <- tryget(my_list$Assembly$Component$.attrs["chromosome"])
        if (is.null(my_chr)) {
            my_chr <- NA
            warning("No chromosomal information for ", SNPs[i], 
                "; may be unmapped")
        }
        my_snp <- tryget(my_list$.attrs["rsId"])
        if (!is.na(my_snp)) {
            my_snp <- paste(sep = "", "rs", my_snp)
        }
        if (my_snp != SNPs[i]) {
            warning(SNPs[i], " has been merged into ", my_snp)
        }
        my_snpClass <- tryget(my_list$.attrs["snpClass"])
        my_gene <- tryget(my_list$Assembly$Component$MapLoc$FxnSet["symbol"])
        if (is.null(my_gene)) 
            my_gene <- NA
        alleles <- my_list$Ss$Sequence$Observed
        if (my_snpClass %in% c("snp", "snv")) {
            tmp <- c(my_list$Ss$Sequence$Observed, my_list$Ss$.attrs["orient"])
            if (tmp[2] != "forward") {
                tmp[1] <- flip(tmp[1], sep = "/")
            }
            alleles_split <- unlist(strsplit(tmp[1], "/"))
            maf_allele <- my_list$Frequency["allele"]
            if (all(maf_allele != alleles_split)) {
                maf_allele <- swap(maf_allele, c("A", "C", "G", 
                  "T"), c("T", "G", "C", "A"))
            }
            if (my_list$Frequency %>% is.null) maf_allele = NA

            my_minor <- alleles_split[maf_allele == alleles_split]
            my_major <- alleles_split[maf_allele != alleles_split]
            my_freq <- my_list$Frequency["freq"]
            if (my_list$Frequency %>% is.null) my_freq = NA
        }
        else {
            my_minor <- NA
            my_major <- NA
            my_freq <- NA
        }
        my_pos <- tryCatch(my_list$Assembly$Component$MapLoc$.attrs["physMapInt"], 
            error = function(e) {
                my_list$Assembly$Component$MapLoc["physMapInt"]
            })
        if (is.null(my_pos)) 
            my_pos <- NA
        out[i, ] <- c(SNPs[i], as.integer(my_chr), my_snp, unname(my_snpClass), 
            unname(my_gene), unname(alleles), unname(my_major), 
            unname(my_minor), as.numeric(my_freq), as.integer(my_pos))
    }
    for (nm in c("MAF", "BP")) {
        out[, nm] <- as.numeric(out[, nm])
    }
    return(out[complete.cases(out),])
    Sys.sleep(3)
}

gsnp(c("rs376101", "rs4820535", "rs3171722", "rs9623964", "rs2066998", "rs2066997"))
```




```{r}
SNPinL5 <- GetSNPsInGenes(GetIDs("IGLL5[SYM]"))
SNPinA1A3 <- GetSNPsInGenes(GetIDs("CACNA1C[SYM]"))

# snp <- GetSNPInfo(c("rs9848", "rs2298845"))
# 
# LDSearch("rs377513170")
sum(complete.cases(getsnp(SNPinA1A3[1:2])))
kable(snp, align = "c")
```

```{r}
library(devtools)
install_github("smbache/magrittr")
library(magrittr)
library(dplyr)

iris %>%
  filter(Species == "virginica") %>%
  select(-Species) %>%
  colMeans
1:10 %>% plot(I(.), col = .) 
1:10 %>% plot(I(.), col = .)
x <- 1:10
x :=replace(1:5, 0)
x2 <- x %>%
  replace(1:5, 0) %>%
  divide_by(2)
```






